cmd/maya-exporter/app/collector/cstorcollector.go

```diff
@@ -0,0 +1,310 @@
+		// the value of net.Conn in case of error so that new connection be created
+		// after the new request from prometheus comes.
+		if err := c.collect(); err != nil {
+			c.Metrics.connectionErrorCounter.WithLabelValues(err.Error()).Inc()
+			glog.Error("Error in connection, closing the connection")
+			c.Conn = nil
+			return err
+		}
+	}
+	return nil
+}
+
+// writer writes the "IOSTATS\n" command on the socket. it returns
+// error if the connection is closed or not available with socket.
+func (c *CstorStatsExporter) writer() error {
+	msg := Command + "\n"
+	_, err := c.Conn.Write([]byte(msg))
+	if err != nil {
+		glog.Error("Write error:", err)
+		return err
+	}
+	return nil
+}
+
+// reader reads the response from socket in the buffer and
+// exits if the buffer contains the "IOSTATS\r\n" at the end.
+func (c *CstorStatsExporter) reader() (string, error) {
+	buf := make([]byte, 256)
+	var (
+		err           error
+		n             int
+		str, response string
+		buffer        bytes.Buffer
+	)
+	// infinite for loop to collect all the chunks.
+	for {
+		n, err = c.Conn.Read(buf[:])
+		if err != nil {
+			glog.Error("Error in reading response, found error : ", err)
+			return "", err
+		}
+		// concatnat the chunks received and then compare if it has
+		// Footer (" IOSTATS") + EOF ("\r\n") and exit, else continue
+		// appending the chunks at the end of str.
+		buffer.WriteString(string(buf[0:n]))
+		str = buffer.String()
+		if str[len(str)-12:] == Footer+EOF {
+			response = str
+			break
+		}
+	}
+	return response, nil
+}
+
+// splitter extracts the JSON from the response :
+// "IOSTATS  { \"iqn\": \"iqn.2017-08.OpenEBS.cstor:vol1\",
+//	\"writes\": \"0\", \"reads\": \"0\", \"totalwritebytes\": \"0\",
+//  \"totalreadbytes\": \"0\", \"size\": \"10737418240\" }\r\nOK IOSTATS\r\n"
+func splitter(resp string) string {
+	var result []string
+	result = strings.Split(resp, EOF)
+	result = v1.Remove(result, Footer)
+	res := strings.TrimPrefix(result[0], Command+"  ")
+	return res
+}
+
+// unmarshaller unmarshal the JSON into Response instances.
+func unmarshaller(result string) Response {
+	metrics := Response{}
+	if err := json.Unmarshal([]byte(result), &metrics); err != nil {
+		glog.Error("Error in unmarshalling, found error: ", err)
+	}
+	glog.Infof("Parsed metrics : %+v", metrics)
+	return metrics
+}
+
+// collect make call to reader and writer to write the
+// IOSTATS command over wire and then reads the response.
+// There are two consecutive calls for read and write over
+// a gap of 1 second to get the I/O stats per second.
+func (c *CstorStatsExporter) collect() error {
+	var (
+		initialResponse, finalResponse string
+		initialMetrics, finalMetrics   Response
+		metrics                        MetricsDiff
+		err                            error
+	)
+	if err := c.writer(); err != nil {
+		return err
+	}
+	initialResponse, err = c.reader()
+	if err != nil {
+		return err
+	}
+	time.Sleep(1 * time.Second)
```

Comment:
This is part of logic. i.e, we get the no of reads/writes etc in one rest api call or over wire, that's why i  do sleep of 1sec so that i can calculate reads/writes per second. While the intervals for prometheus is different. i.e, if prometheus makes call every 5 second then we have to send these read/write per second per 5 second interval, note that this 5 second is done dynamically and controlled from prometheus's end in the configmap.

Reply:
This logic of calculating IOPS and throughput needs to be enhanced. With the current logic, we are calculating IOPs values for 1 sec AFTER receiving the request from the Prom agent. Even though the agent requests every scrape interval, the IOs that have happened between the 2 scrapes are not considered for the IOPs. This needs to be modified. There could be 2 approaches:

1. Make the exporter dumb i.e. let it send the IO metric instead of IOPS, let the client take care of forming the logic for calculating IOPS from the successive IO values.
2. Exporter should store the prev values, fetched during the previous scrape. The difference between the current and prev scrape values can be used to calculate the IOPs.

Reply:
@avishnu this is a good approach and i want to work on this in another PR as a part of refactoring where hardening of things will be done.